### ISSUE TITLE
test_jabbercomponent, test_jabbererror, test_jabberxmlstream, test_msn fail if pyopenssl is not installed.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1881]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| spiv reported|
|-|-|
|Trac ID|trac#1881|
|Type|defect|
|Created|2006-06-30 06:20:06Z|

From http://twistedmatrix.com/buildbot/full-2.5-suse/builds/4/step-shell/2:

```
===============================================================================
[ERROR]: twisted.words.test.test_jabbercomponent

Traceback (most recent call last):
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/trial/runner.py", line 343, in loadByName
    thing = self.findByName(name)
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/trial/runner.py", line 252, in findByName
    return reflect.namedAny(name)
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/python/reflect.py", line 352, in namedAny
    topLevelPackage = __import__(trialname)
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/test/test_jabbercomponent.py", line 8, in <module>
    from twisted.words.protocols.jabber.component import ConnectComponentAuthenticator
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/protocols/jabber/component.py", line 10, in <module>
    from twisted.words.protocols.jabber import jstrports, xmlstream
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/protocols/jabber/xmlstream.py", line 13, in <module>
    from OpenSSL import SSL
exceptions.ImportError: No module named OpenSSL
===============================================================================
[ERROR]: twisted.words.test.test_jabbererror.StreamErrorTest.testGetElementPlain

Traceback (most recent call last):
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/test/test_jabbererror.py", line 47, in testGetElementPlain
    element = e.getElement()
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/protocols/jabber/error.py", line 90, in getElement
    from twisted.words.protocols.jabber.xmlstream import NS_STREAMS
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/protocols/jabber/xmlstream.py", line 13, in <module>
    from OpenSSL import SSL
exceptions.ImportError: No module named OpenSSL
===============================================================================
[ERROR]: twisted.words.test.test_jabberxmlstream

Traceback (most recent call last):
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/trial/runner.py", line 343, in loadByName
    thing = self.findByName(name)
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/trial/runner.py", line 252, in findByName
    return reflect.namedAny(name)
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/python/reflect.py", line 352, in namedAny
    topLevelPackage = __import__(trialname)
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/test/test_jabberxmlstream.py", line 3, in <module>
    from twisted.words.protocols.jabber import error, xmlstream
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/protocols/jabber/xmlstream.py", line 13, in <module>
    from OpenSSL import SSL
exceptions.ImportError: No module named OpenSSL
===============================================================================
[ERROR]: twisted.words.test.test_msn

Traceback (most recent call last):
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/trial/runner.py", line 343, in loadByName
    thing = self.findByName(name)
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/trial/runner.py", line 252, in findByName
    return reflect.namedAny(name)
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/python/reflect.py", line 352, in namedAny
    topLevelPackage = __import__(trialname)
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/test/test_msn.py", line 24, in <module>
    from twisted.words.protocols import msn
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/words/protocols/msn.py", line 85, in <module>
    from twisted.internet.ssl import ClientContextFactory
  File "/home/scmikes/twistedbuildbot/bot-scmikes-2.5/Twisted/twisted/internet/ssl.py", line 47, in <module>
    from OpenSSL import SSL
exceptions.ImportError: No module named OpenSSL
-------------------------------------------------------------------------------
```

At worst, the tests should be skipped, rather than failing.  

Even better, only tests that actually need the OpenSSL module should be skipped, and the others should still run.

<details><summary>Searchable metadata</summary>

```
trac-id__1881 1881
type__defect defect
reporter__spiv spiv
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__words words
keywords__ 
time__1151648406000000 1151648406000000
changetime__1221142224000000 1221142224000000
version__None None
owner__ 
cc__spiv cc__ralphm cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm commented|
|-|-|

`twisted.words.protocols.jabber.xmlstream` imports the `twisted.internet.ssl` module and I think it used to be proper to first try importing OpenSSL.SSL.

I see the 'proper' procedure for importing the ssl module is now somewhat more involved, and I can update the xmlstream module to match. I could also only import this when trying to use the TLS. Would that be better?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

I am not going to fix this until there is a buildslave set up to test the behavior of Twisted in the absence of PyOpenSSL.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

Apparently someone else fixed the jabber issues in some other ticket.  The msn issues are in progress in [#3111](https://github.com/twisted/twisted/issues/3111).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm commented|
|-|-|

Ah yes, this was fixed for the Jabber bits as part of [#2171](https://github.com/twisted/twisted/issues/2171).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
